evaluate_segmentation.py

```diff
@@ -0,0 +1,179 @@
+
+import numpy as np
+import pandas as pd
+import rasterio
+from hydra.utils import get_original_cwd
+from mlflow import log_metrics
+from shapely.geometry import Polygon
+from tqdm import tqdm
+import geopandas as gpd
+
+from utils.geoutils import clip_raster_with_gpkg, vector_to_raster
+from utils.metrics import ComputePixelMetrics
+from utils.utils import get_key_def, list_input_images, get_logger, read_modalities
+from utils.verifications import validate_num_classes, assert_crs_match
+
+logging = get_logger(__name__)
+
+
+def metrics_per_tile(label_arr: np.ndarray, pred_img: np.ndarray, input_image: rasterio.DatasetReader,
+                     chunk_size: int, gpkg_name: str, num_classes: int) -> gpd.GeoDataFrame:
+    """
+    Compute metrics for each tile processed during inference
+    @param label_arr: numpy array of label
+    @param pred_img: numpy array of prediction
+    @param input_image: Rasterio file handle holding the (already opened) input raster
+    @param chunk_size: tile size for per-tile metrics
+    @param gpkg_name: name of geopackage
+    @param num_classes: number of classes
+    @return:
+    """
+    xmin, ymin, xmax, ymax = input_image.bounds  # left, bottom, right, top
+    xres, yres = (abs(input_image.transform.a), abs(input_image.transform.e))
+    mx = chunk_size * xres
+    my = chunk_size * yres
+    h, w = input_image.shape
+
+    feature = defaultdict(list)
+    cnt = 0
+    for row in tqdm(range(0, h, chunk_size), position=2, leave=False):
+        for col in tqdm(range(0, w, chunk_size), position=3, leave=False):
+            label = label_arr[row:row + chunk_size, col:col + chunk_size]
+            pred = pred_img[row:row + chunk_size, col:col + chunk_size]
+            pixelMetrics = ComputePixelMetrics(label.flatten(), pred.flatten(), num_classes)
+            eval = pixelMetrics.update(pixelMetrics.iou)
+            feature['id_image'].append(gpkg_name)
+            for c_num in range(num_classes):
+                feature['L_count_' + str(c_num)].append(int(np.count_nonzero(label == c_num)))
+                feature['P_count_' + str(c_num)].append(int(np.count_nonzero(pred == c_num)))
+                feature['IoU_' + str(c_num)].append(eval['iou_' + str(c_num)])
+            feature['mIoU'].append(eval['macro_avg_iou'])
+            logging.debug(eval['macro_avg_iou'])
+            x_1, y_1 = (xmin + (col * xres)), (ymax - (row * yres))
+            x_2, y_2 = (xmin + ((col * xres) + mx)), y_1
+            x_3, y_3 = x_2, (ymax - ((row * yres) + my))
+            x_4, y_4 = x_1, y_3
+            geom = Polygon([(x_1, y_1), (x_2, y_2), (x_3, y_3), (x_4, y_4)])
+            feature['geometry'].append(geom)
+            feature['length'].append(geom.length)
+            feature['pointx'].append(geom.centroid.x)
+            feature['pointy'].append(geom.centroid.y)
+            feature['area'].append(geom.area)
+            cnt += 1
+    gdf = gpd.GeoDataFrame(feature, crs=input_image.crs.to_epsg())
```

Comment:
@victorlazio109 did you see my secret recipe to get the crs to work? :)
I think you had been struggling with this. I do 2 things:
1. set the epsg of each "inference" benchmark to the epsg of the original inference (i.e. of the imagery). The "to_epsg()" may have been something you were forgetting...
2. set all the benchmarking metrics to epsg 4326 if the inference span across Canada. (see comment below)

Reply:
I think we had solved this issue previously, thanks to you!